File: rust/features/steps/rich_text_signal_steps.rs
//! Step definitions for rich text quality signal scenarios.

use cucumber::{then, when};

use kanbus::cli::run_from_args_with_output;
use kanbus::file_io::load_project_directory;
use kanbus::issue_files::read_issue_from_file;

use crate::step_definitions::initialization_steps::KanbusWorld;

fn run_and_capture(world: &mut KanbusWorld, args: Vec<String>) {
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");

    match run_from_args_with_output(args, cwd.as_path()) {
        Ok(output) => {
            world.exit_code = Some(0);
            world.stdout = Some(output.stdout);
            world.stderr = Some(output.stderr);
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stdout = Some(String::new());
            world.stderr = Some(error.to_string());
        }
    }
}

#[when(expr = "I create an issue with a literal backslash-n description {string}")]
fn when_create_with_literal_backslash_n_description(world: &mut KanbusWorld, description: String) {
    run_and_capture(
        world,
        vec![
            "kanbus".to_string(),
            "create".to_string(),
            "Test".to_string(),
            "Issue".to_string(),
            "--description".to_string(),
            description,
        ],
    );
}

#[when(expr = "I create an issue with a plain-text description {string}")]
fn when_create_with_plain_text_description(world: &mut KanbusWorld, description: String) {
    run_and_capture(
        world,
        vec![
            "kanbus".to_string(),
            "create".to_string(),
            "Test".to_string(),
            "Issue".to_string(),
            "--description".to_string(),
            description,
        ],
    );
}

#[when("I create an issue with a clean multi-line description")]
fn when_create_with_clean_multi_line_description(world: &mut KanbusWorld) {
    let description = "First line\nSecond line\nThird line".to_string();
    run_and_capture(
        world,
        vec![
            "kanbus".to_string(),
            "create".to_string(),
            "Test".to_string(),
            "Issue".to_string(),
            "--description".to_string(),
            description,
        ],
    );
}

#[when(expr = "I comment on {string} with literal backslash-n text {string}")]
fn when_comment_with_literal_backslash_n_text(
    world: &mut KanbusWorld,
    identifier: String,
    text: String,
) {
    std::env::set_var("KANBUS_USER", "dev@example.com");
    run_and_capture(
        world,
        vec![
            "kanbus".to_string(),
            "comment".to_string(),
            identifier,
            text,
        ],
    );
}

#[when(expr = "I comment on {string} with plain text {string}")]
fn when_comment_with_plain_text(world: &mut KanbusWorld, identifier: String, text: String) {
    std::env::set_var("KANBUS_USER", "dev@example.com");
    run_and_capture(
        world,
        vec![
            "kanbus".to_string(),
            "comment".to_string(),
            identifier,
            text,
        ],
    );
}

#[when(expr = "I update {string} with plain-text description {string}")]
fn when_update_with_plain_text_description(
    world: &mut KanbusWorld,
    identifier: String,
    description: String,
) {
    run_and_capture(
        world,
        vec![
            "kanbus".to_string(),
            "update".to_string(),
            identifier,
            "--description".to_string(),
            description,
        ],
    );
}

#[then("the stored description contains real newlines")]
fn then_stored_description_contains_real_newlines(world: &mut KanbusWorld) {
    let cwd = world
        .working_directory
        .as_ref()
        .expect("working directory not set");
    let project_dir = load_project_directory(cwd).expect("project dir");
    let issues_dir = project_dir.join("issues");

    let mut issue_files: Vec<_> = std::fs::read_dir(&issues_dir)
        .expect("read issues dir")
        .filter_map(|entry| entry.ok())
        .filter(|entry| {
            entry
                .path()
                .extension()
                .and_then(|ext| ext.to_str())
                == Some("json")
        })
        .collect();

    issue_files.sort_by_key(|entry| {
        entry.metadata().and_then(|m| m.modified()).ok()
    });

    let issue_path = issue_files
        .last()
        .expect("at least one issue file")
        .path();
    let issue = read_issue_from_file(&issue_path).expect("read issue");

    assert!(
        issue.description.contains('\n'),
        "Expected real newlines in description, got: {:?}",
        issue.description
    );
    assert!(
        !issue.description.contains("\\n"),
        "Expected no literal backslash-n sequences in description, got: {:?}",
        issue.description
    );
}

File: python/features/steps/rich_text_signal_steps.py
"""Behave steps for rich text quality signal scenarios."""

from __future__ import annotations

import re

from behave import then, when

from features.steps.shared import (
    load_project_directory,
    read_issue_file,
    run_cli,
    run_cli_args,
    capture_issue_identifier,
)


_ANSI_RE = re.compile(r"\x1b\[[0-9;]*m")


def _strip_ansi(text: str) -> str:
    return _ANSI_RE.sub("", text)


@when('I create an issue with a literal backslash-n description "{description}"')
def when_create_issue_with_literal_backslash_n_description(
    context: object, description: str
) -> None:
    """Create an issue with a description containing literal \\n sequences.

    :param context: Behave context object.
    :type context: object
    :param description: Description text with literal backslash-n characters.
    :type description: str
    """
    run_cli(context, f"kanbus create Test Issue --description {description!r}")


@when('I create an issue with a plain-text description "{description}"')
def when_create_issue_with_plain_text_description(
    context: object, description: str
) -> None:
    """Create an issue with a plain text description containing no Markdown.

    :param context: Behave context object.
    :type context: object
    :param description: Plain text description.
    :type description: str
    """
    run_cli(context, f"kanbus create Test Issue --description {description!r}")


@when("I create an issue with a clean multi-line description")
def when_create_issue_with_clean_multi_line_description(context: object) -> None:
    """Create an issue with a real multi-line description (no literal \\n sequences).

    Passes the description as a list to avoid shell escaping issues.

    :param context: Behave context object.
    :type context: object
    """
    description = "First line\nSecond line\nThird line"
    run_cli_args(context, ["create", "Test", "Issue", "--description", description])


@when('I comment on "{identifier}" with literal backslash-n text "{text}"')
def when_comment_with_literal_backslash_n_text(
    context: object, identifier: str, text: str
) -> None:
    """Add a comment containing literal backslash-n sequences.

    :param context: Behave context object.
    :type context: object
    :param identifier: Issue identifier.
    :type identifier: str
    :param text: Comment text with literal backslash-n characters.
    :type text: str
    """
    overrides = dict(getattr(context, "environment_overrides", {}) or {})
    overrides["KANBUS_USER"] = "dev@example.com"
    context.environment_overrides = overrides
    run_cli(context, f"kanbus comment {identifier} {text!r}")


@when('I comment on "{identifier}" with plain text "{text}"')
def when_comment_with_plain_text(context: object, identifier: str, text: str) -> None:
    """Add a plain text comment (no Markdown formatting).

    :param context: Behave context object.
    :type context: object
    :param identifier: Issue identifier.
    :type identifier: str
    :param text: Plain comment text.
    :type text: str
    """
    overrides = dict(getattr(context, "environment_overrides", {}) or {})
    overrides["KANBUS_USER"] = "dev@example.com"
    context.environment_overrides = overrides
    run_cli(context, f"kanbus comment {identifier} {text!r}")


@when('I update "{identifier}" with plain-text description "{description}"')
def when_update_issue_with_plain_text_description(
    context: object, identifier: str, description: str
) -> None:
    """Update an issue with a plain-text description.

    :param context: Behave context object.
    :type context: object
    :param identifier: Issue identifier.
    :type identifier: str
    :param description: Plain text description.
    :type description: str
    """
    run_cli(context, f"kanbus update {identifier} --description {description!r}")


@then("the stored description contains real newlines")
def then_stored_description_contains_real_newlines(context: object) -> None:
    """Assert the last created issue has real newlines in its description.

    :param context: Behave context object.
    :type context: object
    """
    identifier = capture_issue_identifier(context)
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert (
        "\n" in issue.description
    ), f"Expected real newlines in description, got: {issue.description!r}"
    assert (
        "\\n" not in issue.description
    ), f"Expected no literal backslash-n sequences in description, got: {issue.description!r}"


@then('stderr should not contain "{text}"')
def then_stderr_not_contains_text(context: object, text: str) -> None:
    """Assert that stderr does not contain the given text.

    :param context: Behave context object.
    :type context: object
    :param text: Text that should not appear in stderr.
    :type text: str
    """
    normalized = text.replace('\\"', '"')
    stderr = _strip_ansi(context.result.stderr)
    assert normalized not in stderr, (
        f"Expected stderr NOT to contain '{normalized}', but it did.\n"
        f"STDERR:\n{stderr}"
    )
